src/Umbraco.Core/Auditing/AuditEventHandler.cs

```diff
@@ -0,0 +1,280 @@
+using Umbraco.Core.Security;
+using Umbraco.Core.Services;
+
+namespace Umbraco.Core.Auditing
+{
+    internal class AuditEventHandler : ApplicationEventHandler
+    {
+        // yada we should inject these
+        private IAuditService AuditServiceInstance => ApplicationContext.Current.Services.AuditService;
+        private IUserService UserServiceInstance => ApplicationContext.Current.Services.UserService;
+
+        private IUser PerformingUser
+        {
+            get
+            {
+                var identity = Thread.CurrentPrincipal?.GetUmbracoIdentity();
+                return identity == null
+                    ? new User { Id = 0, Name = "(no user)", Email = "" }
+                    : UserServiceInstance.GetUserById(Convert.ToInt32(identity.Id));
+            }
+        }
+
+        private string PerformingIp
+        {
+            get
+            {
+                var httpContext = HttpContext.Current == null ? (HttpContextBase) null : new HttpContextWrapper(HttpContext.Current);
+                var ip = httpContext.GetCurrentRequestIpAddress();
+                if (ip.ToLowerInvariant().StartsWith("unknown")) ip = "";
+                return ip;
+            }
+        }
+
+        protected override void ApplicationStarted(UmbracoApplicationBase umbracoApplication, ApplicationContext applicationContext)
+        {
+            //BackOfficeUserManager.AccountLocked += ;
+            //BackOfficeUserManager.AccountUnlocked += ;
+            BackOfficeUserManager.ForgotPasswordRequested += OnForgotPasswordRequest;
+            BackOfficeUserManager.ForgotPasswordChangedSuccess += OnForgotPasswordChange;
+            BackOfficeUserManager.LoginFailed += OnLoginFailed;
+            //BackOfficeUserManager.LoginRequiresVerification += ;
+            BackOfficeUserManager.LoginSuccess += OnLoginSuccess;
+            BackOfficeUserManager.LogoutSuccess += OnLogoutSuccess;
+            BackOfficeUserManager.PasswordChanged += OnPasswordChanged;
+            BackOfficeUserManager.PasswordReset += OnPasswordReset;
+            //BackOfficeUserManager.ResetAccessFailedCount += ;
+
+            UserService.SavedUserGroup += OnSavedUserGroup;
+
+            UserService.SavedUser += OnSavedUser;
+            UserService.DeletedUser += OnDeletedUser;
+            UserService.UserGroupPermissionsAssigned += UserGroupPermissionAssigned;
+
+            MemberService.Saved += OnSavedMember;
+            MemberService.Deleted += OnDeletedMember;
+            MemberService.AssignedRoles += OnAssignedRoles;
+            MemberService.RemovedRoles += OnRemovedRoles;
+        }
+
+        private void OnRemovedRoles(IMemberService sender, MemberService.RolesEventArgs args)
+        {
+            var performingUser = PerformingUser;
+            var roles = string.Join(", ", args.Roles);
+            var members = ApplicationContext.Current.Services.MemberService.GetAllMembers(args.MemberIds).ToDictionary(x => x.Id, x => x);
```

Comment:
We don't need to use the singleton here, the `IMemberService` is passed in as the `sender`. Similarly for the below method.

Reply:
indeed, fixing